_posts/en/2016-12-28-delegates-closures.md

```diff
@@ -0,0 +1,299 @@
+
+@end
+```
+
+Then we implement the methods in our class like this :
+
+```Objective-C
+#import "MyClass.h"
+
+@implementation MyClassObjC
+
+- (void)onRequestFailure {
+
+}
+
+- (void)onRequestSuccess {
+
+}
+
+@end
+```
+
+##### Swift
+
+```Swift
+protocol RequesterDelegateSwift {
+    func onRequestFailure()
+    func onRequestSuccess()
+}
+```
+
+```Swift
+class MyClassSwift: UIViewController, RequesterDelegateSwift {
+
+    func onRequestFailure() {
+
+    }
+
+    func onRequestSuccess() {
+    }
+}
+```
+
+So, we have our Class MyClass that inherits the Protocol RequesterDelegate and that implements 2 methods (onRequestSuccess, onRequestFailure). We are going to do a dummy implementation so that you have an idea of how this works : 
+
+##### Objective-C
+
+```Objective-C
+#import "MyClass.h"
+#import "RequestManager.h"
+
+@implementation MyClassObjC
+
+- (void)callWebService {
+    RequestManager* manager = [[RequestManager alloc] init];
+    manager.delegate = self;
+    [manager get:@"http://plop.fr/json"];
+}
+
+- (void)onRequestFailure {
+
+}
+
+- (void)onRequestSuccess {
+
+}
+
+@end
+```
+
+```Objective-C
+@interface RequestManager : NSObject
+{
+
+}
+
+@property (nonatomic, weak) id delegate;
+
+- (void)get:(NSString*)url;
+```
+
+```Objective-C
+#import "RequestManager.h"
+
+@implementation RequestManager
+
+- (void)get:(NSString *)url {
+    //Do the call
+    BOOL requestSucceed = [self isSuccess];
+
+    //After the call
+    if (requestSucceed) {
+        [self.delegate onRequestSuccess];
+    } else {
+        [self.delegate onRequestFailure];
+    }
+}
+
+- (BOOL)isSuccess {
+    return YES;
+}
+
+@end
+```
+
+##### Swift
+
+```Swift
+class MyClassSwift: UIViewController, RequesterDelegateSwift {
+
+    func callWebService() {
+        let manager: RequestManager = RequestManager()
+        manager.delegate = self
+        manager.get(url: "http://plop.fr/json")
+    }
+
+    func onRequestFailure() {
+
+    }
+
+    func onRequestSuccess() {
+    }
+}
+```
+
+```Swift
+class RequestManager {
+
+    var delegate: RequesterDelegateSwift?
+
+    func get(url: String) {
+        //Do the call
+        let requestSucceed: Bool = self.isSuccess()
+
+        //After the call
+        if requestSucceed {
+            self.delegate?.onRequestSuccess()
+        } else {
+            self.delegate?.onRequestFailure()
+        }
+    }
+
+    private func isSuccess() -> Bool {
+        return true
+    }
+}
+```
+
+**But why do all this?**
+
+We have a light dependency between our different objects. Indeed, RequestManager has no idea of the type of the object of its delegate, all that it cares about is to know that the object contains the two methods defined in the protocol in order to be able to call them. So, It's good, I was able to call my webservice, I have my callback methods in which I can handle the different cases, everything seems good to me. Yes, everything is good, it works well and it will cover a lot of cases
```

Comment:
So, it's fine, I was
Yes, everything is well, it works well and it will cover a lot of cases.